op-e2e/system_test.go

```diff
@@ -171,6 +172,52 @@ func TestSystemE2EDencunAtGenesis(t *testing.T) {
 	require.NotNil(t, head.ExcessBlobGas(), "L1 is building dencun blocks since genesis")
 }
 
+// TestSystemE2EDencunAtGenesis tests if L2 finalizes when blobs are present on L1
+func TestSystemE2EDencunAtGenesisWithBlobs(t *testing.T) {
+	InitParallel(t)
+
+	cfg := DefaultSystemConfig(t)
+	//cancun is on from genesis:
+	genesisActivation := uint64(0)
+	cfg.DeployConfig.L1CancunTimeOffset = &genesisActivation // i.e. turn cancun on at genesis time + 0
+
+	sys, err := cfg.Start(t)
+	require.Nil(t, err, "Error starting up system")
+	defer sys.Close()
+
+	// send a blob-containing txn on l1
+	ethPrivKey := sys.cfg.Secrets.Alice
+	txData := transactions.CreateEmptyBlobTx(ethPrivKey, true)
+	txData.ChainID.SetUint64(uint64(sys.cfg.L1ChainIDBig().Uint64())) // convert from big.int chainid to uint256 chainid
+	tx := types.MustSignNewTx(ethPrivKey, types.LatestSignerForChainID(cfg.L1ChainIDBig()), txData)
+	// send blob-containing txn
+	sendCtx, sendCancel := context.WithTimeout(context.Background(), 15*time.Second)
+	defer sendCancel()
+
+	l1Client := sys.Clients["l1"]
+	err = l1Client.SendTransaction(sendCtx, tx)
+	require.NoError(t, err, "Sending L1 empty blob tx")
+	// Wait for transaction on L1
+	_, err = geth.WaitForTransaction(tx.Hash(), l1Client, 30*time.Duration(cfg.DeployConfig.L1BlockTime)*time.Second)
+	require.Nil(t, err, "Waiting for blob tx on L1")
+	// end sending blob-contraining txns on l1
+
+	l2Seq := sys.Clients["sequencer"]
+
+	// as configured in the extra geth lifecycle in testing setup
+	const finalizedDistance = 8
+	// Wait enough time for L1 to finalize and L2 to confirm its data in finalized L1 blocks
+	time.Sleep(time.Duration((finalizedDistance+6)*cfg.DeployConfig.L1BlockTime) * time.Second)
```

Comment:
Can we try a wait for finalized tag? There is a `waitForSafeHead` in ffp which is not exported. You can also modify `WaitForBlock` but make a version which waits for a block to also be safe/finalized.

Reply:
Vocabulary question: what's a finalized tag, is that like a finalized block? I could definitely pull in `waitForSafeHead`, that sounds like what we want. To confirm, it'd look like:
`waitForSafeHead(some_ctx_with_a_long_deadline, block_number_of_some_kind, sys.Clients["l2"])`? (what would `block_number_of_some_kind` be?)

Reply:
Unsafe/Safe/Finalized are descriptors for blocks. On L1, unsafe is the latest block, safe is a block which is likely to be un-reorgable (assuming some network condition stuff), and finalized is a block which cannot be reorged without slashing occurring (b/c >2/3 of stake has voted for it).

On OP, unsafe means direct from sequencer, safe means batch on L1, finalized means that the batch is from a finalized L1 block.

In this case, why are you waiting for 14 L1 blocks? Are you waiting for an L1 origin to show up on L2? Are you waiting for L1 to finalize? Are you waiting for L2 finalize? Sleeps tend to be much flakier than polling in tests, so I want to figure out what we are waiting for here.

Reply:
Ah, I see, I understand safe vs unsafe vs finalized, but didn't understand we had a notion of tagging blocks with these states in the tests. 
This test is adapted from `TestFinalize`, elsewhere in the file, and so just copies the wait logic pulled from there. Happy to change it for `waitForSafeHead` for some number of blocks `n`.

Reply:
I guess https://github.com/ethereum-optimism/optimism/pull/8196 is related to this and can be used to improve the wait logic here? @trianglesphere

Reply:
Yeah, that's my understanding.